invenio_sequencegenerator/models.py

```diff
@@ -0,0 +1,161 @@
+
+    name = db.Column(db.String, primary_key=True)
+    """The identifying name of the sequence."""
+
+    template = db.Column(db.String, unique=True)
+    """The template of the sequence."""
+
+    parent = db.Column(db.ForeignKey(name))
+    """Indicates if the template depends on another one."""
+
+    @classmethod
+    def create(cls, name, template):
+        """Create a new sequence."""
+        placeholders = extract_placeholders(template)
+
+        # Check that {counter} exists and double its brackets
+        if not any(counter_regex.search('{{{0}}}'.format(p))
+                   for p in placeholders):
+            raise InvalidTemplate('No counter placeholder')
+        template = double_counter(template, counter_regex)
+
+        with db.session.begin_nested():
+            # Check if parent sequence exists
+            try:
+                parent = Sequence.query.filter(
+                    Sequence.name.in_(placeholders)
+                ).one_or_none()
+            except MultipleResultsFound:
+                raise InvalidTemplate('More than 1 parents in template')
+
+            seq = cls(
+                name=name,
+                template=template,
+                parent=parent.name if parent else None
+            )
+            db.session.add(seq)
+
+            seq.kwargs = {}  # state for iterator (fixed kwargs)
+
+            return seq
+
+    @classmethod
+    def get(cls, name):
+        """Get the sequence with the given name."""
+        with db.session.begin_nested():
+            seq = Sequence.query.filter_by(name=name).first()
+            if seq is None:
+                raise NoSuchSequence()
+        return seq
+
+    def get_iter(self, **kwargs):
+        """Get the sequence instantiated with given kwargs."""
+        self.kwargs = kwargs
+        return self
+
+    def next(self, **kwargs):
+        """Get the next report number for the given sequence."""
+        try:
+            template = self.template.format(**kwargs)
+        except KeyError:
+            raise InvalidCall('Improper arguments for string formatting')
+
+        return TemplatedCounter.next_or_create(template)
+
+    def reset(self, **kwargs):
+        """Reset the sequence's counter for these particular kwargs."""
+        try:
+            template = self.template.format(**kwargs)
+        except KeyError:
+            raise InvalidCall('Improper arguments for string formatting')
+
+        TemplatedCounter.query.filter_by(template=template).first().reset()
+
+    def __iter__(self):
+        """Provide iterative capabilities to sequences."""
+        return SequenceIterator(self, self.kwargs)
+
+
+class TemplatedCounter(db.Model):
+    """Stores generated report numbers."""
+
+    __tablename__ = 'sequencegenerator_templatedcounter'
+
+    template = db.Column(db.String, nullable=False,
+                         primary_key=True, index=True)
+
+    counter = db.Column(db.Integer, nullable=False)
```

Comment:
The reason here is that you just want to disable the counter updates because you are manually updating it in https://github.com/inveniosoftware/invenio-sequencegenerator/pull/6/files#diff-a46e42a1b84af85c5eb4bae67b8ff8f6R146